lib/safari.js

```diff
@@ -1,53 +1,173 @@
-import { fs, process, logger } from 'appium-support';
+import os from 'os';
+import path from 'path';
+import { JWProxy, errors } from 'appium-base-driver';
+import { fs, logger, util } from 'appium-support';
 import { SubProcess } from 'teen_process';
+import { waitForCondition } from 'asyncbox';
+import { findAPortNotInUse } from 'portscanner';
+import { execSync } from 'child_process';
 
-const log = logger.getLogger('SafariDriverProc');
+const log = logger.getLogger('SafariDriverServer');
 
-class SafariDriverProc {
-  constructor ({port, unexpectedExitHandler}) {
-    this.port = port;
+const SD_BINARY = 'safaridriver';
+const STARTUP_TIMEOUT = 10000; // seconds
+const SAFARI_PORT_RANGE = [5100, 5200];
+// This guard is needed to make sure
+// we never run multiple Safari driver processes for the same Appium process
+const SAFARI_SERVER_GUARD = util.getLockFileGuard(
+  path.resolve(os.tmpdir(), 'safari_server_guard.lock'),
+  {timeout: 5, tryRecovery: true}
+);
+
+
+class SafariProxy extends JWProxy {
+  async proxyCommand (url, method, body = null) {
+    if (this.didProcessExit) {
+      throw new errors.InvalidContextError(
+        `'${method} ${url}' cannot be proxied to Safari Driver server because ` +
+        'the process is not running (probably crashed). Check the server log for more details');
+    }
+    return await super.proxyCommand(url, method, body);
+  }
+}
+
+
+class SafariDriverProcess {
+  constructor () {
+    this.port = null;
     this.proc = null;
-    this.unexpectedExitHandler = unexpectedExitHandler;
   }
 
-  async start () {
-    await process.killProcess('safaridriver');
-    const safariBin = await fs.which('safaridriver');
-    this.proc = new SubProcess(safariBin, ['-p', this.port]);
+  get isRunning () {
+    return !!(this.proc?.isRunning);
+  }
+
+  async init () {
+    await SAFARI_SERVER_GUARD(async () => {
+      if (this.isRunning) {
+        return;
+      }
+
+      const [startPort, endPort] = SAFARI_PORT_RANGE;
+      try {
+        this.port = await findAPortNotInUse(startPort, endPort);
+      } catch (e) {
+        throw new Error(
+          `Cannot find any free port in range ${startPort}..${endPort}}. ` +
+          `Double check the processes that are locking ports within this range and terminate ` +
+          `these which are not needed anymore`);
+      }
+
+      let safariBin;
+      try {
+        safariBin = await fs.which(SD_BINARY);
```

Comment:
Will we allow to set a custom path for Safari Preview?
On Safari preview, the package includes a safaridriver for the preview version. Then, this path could be like `/Applications/Safari Preview/xxxx/yyy/safaridriver`. Then, we should define a custom path in our capability and allow here customisable.
(As this PR, hardcoded to default one is pretty fine for me)

Reply:
in such case one could just simply prepend the desired root to PATH, so then that binary will be used instead of the one in /usr/bin

Reply:
good idea